src/hierarchies/lin_reg_uni_hierarchy.cpp

```diff
@@ -0,0 +1,192 @@
+#include "lin_reg_uni_hierarchy.hpp"
+
+#include <Eigen/Dense>
+#include <stan/math/prim/err.hpp>
+
+#include "../utils/eigen_utils.hpp"
+#include "../utils/proto_utils.hpp"
+#include "../utils/rng.hpp"
+
+void LinRegUniHierarchy::initialize() {
+  assert(prior != nullptr && "Error: prior was not provided");
+  state.regression_coeffs = hypers->mean;
+  state.var = hypers->scale / (hypers->shape + 1);
+  clear_data();
+}
+
+void LinRegUniHierarchy::clear_data() {
+  mixed_prod = Eigen::VectorXd::Zero(dim);
+  data_sum_squares = 0.0;
+  covar_sum_squares = Eigen::MatrixXd::Zero(dim, dim);
+  card = 0;
+  cluster_data_idx = std::set<int>();
+}
+
+void LinRegUniHierarchy::update_summary_statistics(
+    const Eigen::VectorXd &datum, const Eigen::VectorXd &covariate, bool add) {
+  if (add) {
+    data_sum_squares += datum(0) * datum(0);
+    covar_sum_squares += covariate * covariate.transpose();
+    mixed_prod += datum(0) * covariate;
+  } else {
+    data_sum_squares -= datum(0) * datum(0);
+    covar_sum_squares -= covariate * covariate.transpose();
+    mixed_prod -= datum(0) * covariate;
+  }
+}
+
+LinRegUniHierarchy::Hyperparams LinRegUniHierarchy::normal_invgamma_update() {
+  Hyperparams post_params;
+
+  post_params.var_scaling = covar_sum_squares + hypers->var_scaling;
+  post_params.mean = stan::math::inverse_spd(post_params.var_scaling) *
+                     (mixed_prod + hypers->var_scaling * hypers->mean);
+  post_params.shape = hypers->shape + 0.5 * card;
+  post_params.scale =
+      hypers->scale +
+      0.5 * (data_sum_squares +
+             hypers->mean.transpose() * hypers->var_scaling * hypers->mean -
+             post_params.mean.transpose() * post_params.var_scaling *
+                 post_params.mean);
+  return post_params;
+}
+
+void LinRegUniHierarchy::update_hypers(
+    const std::vector<bayesmix::MarginalState::ClusterState> &states) {
+  auto &rng = bayesmix::Rng::Instance().get();
+  if (prior->has_fixed_values()) {
+    return;
+  }
+
+  else {
+    throw std::invalid_argument("Error: unrecognized prior");
+  }
+}
+
+double LinRegUniHierarchy::like_lpdf(
+    const Eigen::RowVectorXd &datum,
+    const Eigen::RowVectorXd &covariate) const {
+  return stan::math::normal_lpdf(
+      datum(0), state.regression_coeffs.dot(covariate), sqrt(state.var));
+}
+
+Eigen::VectorXd LinRegUniHierarchy::like_lpdf_grid(
+    const Eigen::MatrixXd &data, const Eigen::MatrixXd &covariates) const {
+  Eigen::VectorXd result(data.rows());
+  for (size_t i = 0; i < data.rows(); i++) {
+    result(i) = like_lpdf(data.row(i), covariates.row(i));
+  }
+  return result;
+}
+
+double LinRegUniHierarchy::marg_lpdf(
+    const Eigen::RowVectorXd &datum,
+    const Eigen::RowVectorXd &covariate) const {
+  double sig_n =
+      sqrt((1 + (covariate * stan::math::inverse_spd(hypers->var_scaling) *
```

Comment:
since `hypers->var_scaling` is fixed, or updated at most once per iteration, wile `marg_lpdf` can be called thousands of time per iteration, I would save the inverse of this matrix in the `hypers` directly.
Also `var_scaling` can be fairly large, so that would save a lot of time